### PR TITLE
Azure aws actor

### PR DESCRIPTION
**Pull request by PradeepKiruvale**
## Proposed changes

Port the existing azure and aws mappers to actor based mappers

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue
https://github.com/thin-edge/thin-edge.io/issues/1825
## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [x] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: crates/core/tedge_api/Cargo.toml [hunk 1]
```diff
@@ -25,9 +26,10 @@ time = { version = "0.3", features = [
     "serde",
     "serde-well-known",
 ] }
+tracing = { version = "0.1", features = ["attributes", "log"] }
 
 [dev-dependencies]
-anyhow = "1.0"
+anyhow = "1.0.6"
```
[the remark this answers is not in the document]
Comment (PradeepKiruvale) [reply crates/core/tedge_api/Cargo.toml:32]: fixed

File: crates/extensions/tedge_mapper_core/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+async-trait = "0.1"
+c8y_api = { path = "../../core/c8y_api" }
+clock = { path = "../../common/clock" }
+csv = "1.1"
+download = { path = "../../common/download" }
+flockfile = { path = "../../common/flockfile" }
+futures = "0.3"
+json-writer = { path = "../../common/json_writer" }
+logged_command = { path = "../../common/logged_command" }
+mockall = "0.11"
```
[the remark this answers is not in the document]
Comment (PradeepKiruvale) [reply crates/extensions/tedge_mapper_core/Cargo.toml:23]: Not really, wonder why the `cargo udeps` dint catch this.

File: crates/extensions/aws_mapper_ext/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+async-trait = "0.1"
+c8y_api = { path = "../../core/c8y_api" }
+clock = { path = "../../common/clock" }
+log = "0.4"
+serde = { version = "1.0", features = ["derive"] }
+serde_json = "1.0"
+tedge_actors = { path = "../../core/tedge_actors" }
+tedge_api = { path = "../../core/tedge_api" }
+tedge_config = { path = "../../common/tedge_config" }
+tedge_health_ext = { path = "../../extensions/tedge_health_ext" }
```
[the remark this answers is not in the document]
Comment (didier-wenzek) [reply crates/extensions/aws_mapper_ext/Cargo.toml:23]: To be fair, I made the same mistake for the collectd mapper. See https://github.com/thin-edge/thin-edge.io/pull/1817#discussion_r1142148938.

The `run` method should be provided by the `tedge-mapper` crate and not by each and every specific mapper actor.

**Review by albinsuresh — COMMENTED** [review]
A few questions that came to my mind while reviewing the PR:

1. Do we really need to preserve the existing architecture of the mapper binary where every mapper is defined in a common place and always bundle them together? I'm really against this current packaging philosophy as it leads to tight coupling between the core and new mappers(core needs to change every time a new mapper is added).
2. Are the older abstractions like  `TEdgeComponent` really relevant in the new model? I'd say no, as that abstraction is majorly about Mqtt level contracts which are handled by the MqttActor already. There are some other `init` aspects as well, which I expect to be resolved separately #1764. So, I'd try to get rid of this `component` abstraction completely. An `Actor` is the new `TEdgeComponent` in my view. But, if we're going to keep it, then some of the commonality in the `run` methods of all the mappers could be abstracted out into it.
3. There's scope stripping down the `Converter` trait as well, especially the MQTT bits in it.

File: crates/extensions/aws_mapper_ext/src/mapper.rs [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+        let aws_actor = aws_actor.with_connection(&mut mqtt_actor);
+
+        //Instantiate health monitor actor
+        let health_actor = HealthMonitorBuilder::new(self.session_name());
+        mqtt_actor.mqtt_config = health_actor.set_init_and_last_will(mqtt_actor.mqtt_config);
+        let health_actor = health_actor.with_connection(&mut mqtt_actor);
+
+        runtime.spawn(mqtt_actor).await?;
+        runtime.spawn(aws_actor).await?;
+        runtime.spawn(health_actor).await?;
```
[the remark this answers is not in the document]
Comment (PradeepKiruvale) [reply crates/extensions/aws_mapper_ext/src/mapper.rs:86]: Added

File: crates/extensions/tedge_mapper_core/Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,67 @@
+[package]
+name = "tedge_mapper_core"
```
[the remark this answers is not in the document]
Comment (PradeepKiruvale) [reply crates/extensions/tedge_mapper_core/Cargo.toml:2]: I agree, I need to move it to some familiar place where all the mappers can use these things.

[the remark this answers is not in the document]
Comment (didier-wenzek) [reply crates/extensions/tedge_mapper_core/Cargo.toml:2]: I'm not convinced that we need a crate for that.

**Comment by didier-wenzek** [comment]
> Do we really need to preserve the existing architecture of the mapper binary where every mapper is defined in a common place and always bundle them together? I'm really against this current packaging philosophy as it leads to tight coupling between the core and new mappers(core needs to change every time a new mapper is added).

Yes, as a temporary step. One should not change the user interface (i.e. having a single `tedge-mapper` executable and specialized processes as  `tedge-mapper c8y`). 
 
>    Are the older abstractions like TEdgeComponent really relevant in the new model? I'd say no, as that abstraction is majorly about Mqtt level contracts which are handled by the MqttActor already. There are some other init aspects as well, which I expect to be resolved separately https://github.com/thin-edge/thin-edge.io/issues/1764. So, I'd try to get rid of this component abstraction completely. An Actor is the new TEdgeComponent in my view. But, if we're going to keep it, then some of the commonality in the run methods of all the mappers could be abstracted out into it.

I fully agree that `TEdgeComponent` makes sense *only* within the current `tedge-mapper`. The `aws_mapper_ext` must be independent of that.

>    There's scope for stripping down the Converter trait as well, especially the MQTT bits in it.

We can even have a generic actor that uses a `Converter<Input,Output>`.

>   Although there's nothing wrong with the current approach taken to move all the existing actors into extensions, I was expecting something lighter where the existing az and aws modules under tedge_mapper be actorized in-place by just adding a dependency on the tedge_actors crate rather than moving them out. Once these modules are actorized, they could be just spawned from the existing tedge_mapper/main itself.

Yes, this needs to be kept in mind: as a second step the mapper actors one should be able to spawn them in various main.rs.

**Pull request closed without merging**